### ISSUE TITLE
Offscreen VTK rendering dies on Linux

### ISSUE THREAD

**Issue by remram44** [VisTrails/VisTrails#1082]
Running the example API notebook on Linux:
```
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  156 (GLX)
  Minor opcode of failed request:  3 (X_GLXCreateContext)
  Value in failed request:  0x0
  Serial number of failed request:  30
  Current serial number in output stream:  35
```
Then the whole process dies (VTK calls exit?)

**Comment by remram44** [comment]
Also dies on my Mac:
```
ERROR: In /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_release_tarballs_ports_graphics_vtk5/vtk5/work/VTK5.10.1/Graphics/vtkStructuredPointsGeometryFilter.cxx, line 24
vtkStructuredPointsGeometryFilter (0x7fe548d0ce90): vtkStructuredPointsGeometryFilter will be deprecated in
the next release after VTK 4.0. Please use
vtkImageDataGeometryFilter instead

QWidget: Must construct a QApplication before a QPaintDevice
```

**Comment by remram44** [comment]
@rexissimus Can you try on the Mac binary? Am I the only one getting this?

```sh-session
$ export PYTHONHOME=/Applications/VisTrails\ 2.2.0-dev/VisTrails.app/Contents/Resources
$ export DYLD_LIBRARY_PATH=VisTrails.app/Contents/MacOS/vistrails
$ /Applications/VisTrails 2.2.0-dev$ VisTrails.app/Contents/MacOS/python
Python 2.7.2 (v2.7.2:8527427914a2, Jun 11 2011, 15:22:34) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import vistrails as vt
>>> render = vt.load_vistrail('vistrails/examples/api/brain_output.xml')
>>> render.select_latest_version()
>>> render.execute()
QWidget: Must construct a QApplication before a QPaintDevice
Abort trap: 6
$
```

**Comment by dakoop** [comment]
@remram44, I can reproduce this. Does the API or the vistrail set any flags to let vtkRendererOutput know that it wants a file output? Also, in the past, VTK's Mac offscreen code still needed a window to draw into and would flash up a window before hiding it.

**Comment by rexissimus** [comment]
I get this as well. I think the offscreen code still requires a window. I do not know if that is still needed.

**Comment by remram44** [comment]
Tommy wrote `render_to_image()` (as far as I can see, he added it **inside of the merge commit 426a7755**!?)

OffScreenRenderingOn() gets called but VTK just abort() on `window.Render()`.

**Comment by rexissimus** [comment]
`render_to_image` was created in an `OutputModule` branch but got lost somewhere when we merged with Dave's rewrite of the VTK package. I just added it back.

It probably needs an initialized QApplication? We can try removing the widget creation? If this does not work maybe we can make it create its own temporary QApplication? But this would probably not fix the problem on Linux.

**Comment by remram44** [comment]
I wrote [a simple script](https://gist.github.com/remram44/cc0f20055864722fbbf1) that follows the wiki exactly (http://www.vtk.org/Wiki/VTK/Examples/Cxx/Utilities/OffScreenRendering) but gives me the same crash.

**Comment by rexissimus** [comment]
Your script works for me on Ubuntu with VTK 5.8.0 and VTK 6.0.0. I had to remove vtkImagingFactory on VTK6. It also works for me on Mac. Are you running in a VM?

**Comment by remram44** [comment]
This works for me on Mac (VTK 5.10) though render_to_image() doesn't appear to.

**Comment by remram44** [comment]
However it crashes on Linux, with vtk 5.8.0:
```
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  156 (GLX)
  Minor opcode of failed request:  3 (X_GLXCreateContext)
  Value in failed request:  0x0
  Serial number of failed request:  30
  Current serial number in output stream:  35
```
but works on vtk 6.2.0 (also had to remove vtkImagingFactory). Might be an issue with that particular build...

Does render_to_image() work on your machine, i.e. can you run the [API example](https://github.com/VisTrails/VisTrails/blob/master/examples/api/ipython-notebook.ipynb)?

**Comment by rexissimus** [comment]
I ran `ipython notebook examples/api/ipython-notebook.ipynb` on Ubuntu and executed all cells. The brain example renders correctly on VTK 5.8.0. VTK 6 also works, but there is a `SetInput` bug in `render_to_image()`. But once I fixed that (using `writer.SetInputData`), the the brain renders.

I tried running VisTrails with VTK 5.10 in a VirtualBox but it failed with OpenGL errors. I had to switch to VMware and upgrade to VTK 6 to fix my problems. VTK seems picky on what platform it runs on. I think it even uses VM-specific flags to detect its environment. I would not be surprised if you have a bad build.
